onnxruntime/python/tools/quantization/calibrate.py

```diff
@@ -226,12 +239,242 @@ def compute_range(self):
+        '''
+        :param model: ONNX model to calibrate. It can be a ModelProto or a model path
+        :param op_types_to_calibrate: operator types to calibrate. By default, calibrate all the float32/float16 tensors.
+        :param augmented_model_path: save augmented model to this path.
+        '''
+        super(EntropyCalibrater, self).__init__(model, op_types_to_calibrate, augmented_model_path)
+        self.intermediate_outputs = []
+        self.calibrate_tensors_range = None
+        self.num_model_outputs = len(self.model.graph.output)
+        self.model_original_outputs = set(output.name for output in self.model.graph.output)
+        self.collector = None
+
+    def augment_graph(self):
+        '''
+        make all quantization_candidates op type nodes as part of the graph output.
+        :return: augmented ONNX model
+        '''
+        model = onnx_proto.ModelProto()
+        model.CopyFrom(self.model)
+        model = onnx.shape_inference.infer_shapes(model)
+
+        added_nodes = []
+        added_outputs = []
+        tensors, value_infos = self.select_tensors_to_calibrate(model) 
+
+        for tensor in tensors:
+            added_outputs.append(value_infos[tensor])
+
+        model.graph.node.extend(added_nodes)
+        model.graph.output.extend(added_outputs)
+        onnx.save(model, self.augmented_model_path)
+        self.augment_model = model
+
+    def clear_collected_data(self):
+        self.intermediate_outputs = []
+
+    def collect_data(self, data_reader: CalibrationDataReader):
+        '''
+        Entropy Calibrator collects operators' tensors as well as generates tensor histogram for each operator. 
+        '''
+        while True:
+            inputs = data_reader.get_next()
+            if not inputs:
+                break
+            self.intermediate_outputs.append(self.infer_session.run(None, inputs))
+
+
+        if len(self.intermediate_outputs) == 0:
+            raise ValueError("No data is collected.")
+
+        output_names = [self.infer_session.get_outputs()[i].name for i in range(len(self.intermediate_outputs[0]))]
+        output_dicts_list = [
+            dict(zip(output_names, intermediate_output)) for intermediate_output in self.intermediate_outputs
+        ]
+
+        merged_dict = {}
+        for d in output_dicts_list:
+            for k, v in d.items():
+                merged_dict.setdefault(k, []).append(v)
+
+        clean_merged_dict = dict((i, merged_dict[i]) for i in merged_dict if i not in self.model_original_outputs)
+
+        if not self.collector:
+            self.collector = HistogramCollector()
+        self.collector.collect(clean_merged_dict)
+
+    def compute_range(self):
+        ''' 
+        Compute the min-max range of tensor
+        :return: dictionary mapping: {added node names: (ReduceMin, ReduceMax) pairs }
+        '''
+        if not self.collector:
+            print("No collector created and can't generate calibration data.")
```

Comment:
>print [](start = 12, length = 5)

could you please use logging to keep consistent

Reply:
Done